### PR TITLE
Navigation Block: Use dom.focus for focus control

### LINKED ISSUE #57359: o.querySelector(...) is null in /blocks/navigation/view.min.js
### Description

I added the search block to a separate navigation in the header as shown below in order to open the search modal on click.
header-editor
It works as expected, however when the menu button is clicked and search modal opens - a JS error is thrown in the console:
header-front
This happens with the default Twenty Twenty-Four theme and no active plugins.

### Step-by-step reproduction instructions

1. Add navigation to the default header in Twenty Twenty-Four theme.
2. Add search block to the navigation as shown on the first screenshot.
3. Click the menu button on the frontend to open the search modal.
4. Check browser console.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by t-hamano**
Fixes #57359

## What?

This PR avoids using `querySelector`/`querySelectorAll` and uses `focus` from the `@wordpress/dom` package for focus control in the navigation dialog. This fixes a focus error when the Navigation block has a search block, as reported in #57359.

## Why?

The following two things are expected in the Navigation block dialog:

1. Focus on the first focusable element when opening a dialog
2. Restrict focus movement to the dialog when pressing the Tab key

Regarding the first point, it tries to find focusable elements in the dialog starting from the element with the `.wp-block-navigation-item` class. However, if it is just a Search block, there is no element with this class, so focus will fail and you will get a console error.

## How?

~~The `@wordpress/dom` package has a `focus` object that finds focusable elements. I used this and replaced `querySelector`/`querySelectorAll`. This should always focus the appropriate first element, no matter what blocks the navigation block contains.~~

Update: To avoid the increase in bundle size, I fixed the logic of a problem (searching for the first focusing element in the dialog) without using the `@wordpress/dom` package.
​
I also moved the `focusFirstElement` callback to the content wrapper div (`.wp-block-navigation__responsive-container-content`) to accurately locate the first element (not including the close button) when the dialog opens.

Before:

```html
<nav class="wp-block-navigation>
  <button class="wp-block-navigation__responsive-container-open"></button>
  <div class="wp-block-navigation__responsive-container">
    <div class="wp-block-navigation__responsive-close">
      <div class="wp-block-navigation__responsive-dialog" data-wp-watch="callbacks.focusFirstElement">
        <button aria-label="Close menu" class="wp-block-navigation__responsive-container-close"><</button>
        <div class="wp-block-navigation__responsive-container-content">
          <!-- Blocks -->
        </div>
      </div>
    </div>
  </div>
</nav>
```

After:

```html
<nav class="wp-block-navigation>
  <button class="wp-block-navigation__responsive-container-open"></button>
  <div class="wp-block-navigation__responsive-container">
    <div class="wp-block-navigation__responsive-close">
      <div class="wp-block-navigation__responsive-dialog">
        <button class="wp-block-navigation__responsive-container-close"></button>
        <div class="wp-block-navigation__responsive-container-content" data-wp-watch="callbacks.focusFirstElement">
          <!-- Blocks -->
        </div>
      </div>
    </div>
  </div>
</nav>
```

### Testing Instructions for Keyboard

I'm a Windows user, and I've confirmed it works in Chrome and Firefox browsers. I would be happy if you could confirm that it works as expected on Mac OS.

- Add the following two navigations to the template. In both cases, change the Overlay Menu setting to "Always".
  - Contains only a Search block
  - Contains page list or page links
- Use the `Tab` key to focus on the navigation button and press `Enter`.
  - For navigations with a list of pages, the focus should be on the first page link.
  - If the navigation has only a search block, the focus should be on the text field _inside_ the search block.
- Press the Tab key continuously.
- Focus should not leave the dialog.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/54422211/8f019046-23bb-42a0-8c4f-510c72f59b02

### CODE REVIEW

File: test/e2e/specs/editor/blocks/navigation-frontend-interactivity.spec.js [hunk 1]
```diff
@@ -112,7 +112,11 @@ test.describe( 'Navigation block - Frontend interactivity', () => {
 			// Test: overlay menu focuses on first element after opening
 			await expect( overlayMenuFirstElement ).toBeFocused();
 
-			// Not Tested: overlay menu traps focus
+			// Test: overlay menu traps focus
+			await pageUtils.pressKeys( 'Tab', { times: 2, delay: 50 } );
+			await expect( closeMenuButton ).toBeFocused();
+			await pageUtils.pressKeys( 'Shift+Tab', { times: 2, delay: 50 } );
+			await expect( overlayMenuFirstElement ).toBeFocused();
```
Comment (t-hamano) [inline test/e2e/specs/editor/blocks/navigation-frontend-interactivity.spec.js:119]: I didn't know why this test wasn't done in Safari, but let's try it and see if it works.

**Review by Mamaduka — COMMENTED** [review]
Using the `dom` package could significantly increase the [bundle size](https://bundlephobia.com/package/@wordpress/dom@3.48.0) (for a front-end script). Using plain JS might be better for the moment.

**Comment by t-hamano** [comment]
> Using the `dom` package could significantly increase the [bundle size](https://bundlephobia.com/package/@wordpress/dom@3.48.0) (for a front-end script). Using plain JS might be better for the moment.

Thank you for teaching me. Certainly, I think that it can be modified with a simple code without using this package, so I will try it.

**Review by luisherranz — APPROVED** [review]
Look good to me, both the code and the fix 🙂

File: lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php [hunk 2]
```diff
@@ -428,27 +428,29 @@ private static function get_responsive_container_markup( $attributes, $inner_blo
 		$responsive_dialog_directives    = '';
 		$close_button_directives         = '';
 		if ( $should_load_view_script ) {
-			$open_button_directives          = '
```
Comment (getdave) [inline lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php:431]: I checked and these changes are still included in the `WP_Navigation_Block_Renderer` which has now been moved to the block-library package and thus will be automatically synced to Core as part of the packages sync process.

**Comment by getdave** [comment]
✅ I updated the [PHP Sync Tracking Issue for WP 6.5](https://github.com/WordPress/gutenberg/issues/57959) to note this PR no longer requires a backport for WP 6.5.

Why? Because the code was moved into the block-library package in https://github.com/WordPress/gutenberg/pull/57979 and thus should be automatically handled by the packages sync process.

**Pull request merged**